File: library/src/main/java/de/westnordost/osmfeatures/LocalizedFeature.java
package de.westnordost.osmfeatures;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.Map;

class LocalizedFeature implements Feature {

    private final BaseFeature p;
    private final String name;
    private final List<String> terms;
    private final String canonicalName;
    private final List<String> canonicalTerms;
    private final Locale locale;

    public LocalizedFeature(BaseFeature p, Locale locale, String name, List<String> terms)
    {
        this.p = p;
        this.name = name;
        this.terms = terms;
        this.locale = locale;

        this.canonicalName = StringUtils.canonicalize(name);
        List<String> canonicalTerms = new ArrayList<>(terms.size());
        for (String term : terms)
        {
            canonicalTerms.add(StringUtils.canonicalize(term));
        }
        this.canonicalTerms = Collections.unmodifiableList(canonicalTerms);
    }

    @Override public String getId() { return p.getId(); }
    @Override public Map<String, String> getTags() { return p.getTags(); }
    @Override public List<GeometryType> getGeometry() { return p.getGeometry(); }
    @Override public String getName() { return name; }
    @Override public String getIcon() { return p.getIcon(); }
    @Override public String getImageURL() { return p.getImageURL(); }
    @Override public List<String> getTerms() { return terms; }
    @Override public List<String> getIncludeCountryCodes() { return p.getIncludeCountryCodes(); }
    @Override public List<String> getExcludeCountryCodes() { return p.getExcludeCountryCodes(); }
    @Override public boolean isSearchable() { return p.isSearchable(); }
    @Override public double getMatchScore() { return p.getMatchScore(); }
    @Override public boolean isSuggestion() { return p.isSuggestion(); }
    @Override public Map<String, String> getAddTags() { return p.getAddTags(); }
    @Override public Map<String, String> getRemoveTags() { return p.getRemoveTags(); }
    @Override public String getCanonicalName() { return canonicalName; }
    @Override public List<String> getCanonicalTerms() { return canonicalTerms; }
    @Override public Locale getLocale() { return locale; }
}

File: library/src/commonMain/kotlin/de.westnordost.osmfeatures/LocalizedFeature.kt
package de.westnordost.osmfeatures

/** Data class associated with the Feature interface. Represents a localized feature.
 *
 * I.e. the name and terms are specified in the given locale.  */
class LocalizedFeature(
    private val p: BaseFeature,
    override val locale: Locale?,
    override val names: List<String>,
    override val terms: List<String>
) : Feature {
    override val canonicalNames: List<String> = names.map { name -> StringUtils.canonicalize(name) }
    override val canonicalTerms: List<String> = terms.map { term -> StringUtils.canonicalize(term) }

    override val id: String
        get() = p.id
    override val tags: Map<String, String>
        get() = p.tags
    override val geometry: List<GeometryType>
        get() = p.geometry
    override val name: String
        get() = names[0]
    override val icon: String
        get() = p.icon
    override val imageURL: String
        get() = p.imageURL
    override val includeCountryCodes: List<String>
        get() = p.includeCountryCodes
    override val excludeCountryCodes: List<String>
        get() = p.excludeCountryCodes
    override val isSearchable: Boolean
        get() = p.isSearchable
    override val matchScore: Float
        get() = p.matchScore
    override val addTags: Map<String, String>
        get() = p.addTags
    override val removeTags: Map<String, String>
        get() = p.removeTags
    override val isSuggestion: Boolean
        get() = p.isSuggestion

    override fun toString(): String {
        return id
    }
}
